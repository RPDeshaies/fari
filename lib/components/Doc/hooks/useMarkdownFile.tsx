import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import { showdownConverter } from "../../../constants/showdownConverter";
import { useLogger } from "../../../contexts/InjectionsContext/hooks/useLogger";

export type ILoadFunction = () => Promise<string>;

export type ITableOfContent = Record<
  string,
  {
    page: IMarkdownHeader;
    children: Array<IMarkdownHeader>;
  }
>;

export type IMarkdownHeader = {
  id: string;
  label: string;
  preview: string;
  level: number;
  grouping: string;
  page: IMarkdownHeader | undefined;
};

export function useMarkdownFile(loadFunction: ILoadFunction) {
  const [dom, setDom] = useState<HTMLDivElement>();
  const [html, setHtml] = useState<string | undefined>();
  const [tableOfContent, setTableOfContent] = useState<ITableOfContent>({});
  const [allHeaders, setAllHeaders] = useState<Array<IMarkdownHeader>>([]);
  const logger = useLogger();

  useEffect(() => {
    load();
    async function load() {
      if (loadFunction) {
        try {
          const markdown = await loadFunction();

          if (markdown) {
            const {
              dom,
              headers,
              tableOfContent: newTableOfContent,
            } = Markdown.process(markdown);
            setDom(dom);
            setHtml(dom.innerHTML);
            setAllHeaders(headers);
            setTableOfContent(newTableOfContent);
          }
        } catch (error) {
          logger.error("useMarkdownFile:error", error);
        }
      }
    }
  }, [loadFunction]);

  return { html, tableOfContent, dom, allHeaders };
}

export const scrollMarginTop = 16;
class Markdown {
  public static process(markdown: string) {
    const html = showdownConverter.makeHtml(markdown);
    const dom = document.createElement("div");
    dom.innerHTML = html;

    let tableOfContent: ITableOfContent = {};
    const headers: Array<IMarkdownHeader> = [];

    let latestH1:
      | IMarkdownHeader
      | undefined = (undefined as unknown) as IMarkdownHeader;
    const allHeaders = dom.querySelectorAll("h1,h2,h3,h4,h5,h6");

    for (let i = 0; i < allHeaders.length; i++) {
      const element = allHeaders[i];

      const elementLevel = this.getElementLevel(element);

      if (i === 0 && elementLevel !== 1) {
        throw "Missing top level <h1/> tag";
      }

      const label = element.textContent ?? "";
      const nextElement = element.nextElementSibling;

      const canPreviewNextElement = this.isElementHeader(nextElement);
      const preview = canPreviewNextElement
        ? nextElement?.textContent?.trim() ?? ""
        : "";

      if (elementLevel === 1) {
        const header1: IMarkdownHeader = {
          id: element.id,
          label: label,
          preview: preview,
          level: elementLevel,
          page: undefined,
          grouping: label,
        };

        latestH1 = header1;
        tableOfContent = {
          ...tableOfContent,
          [element.id]: { page: header1, children: [] },
        };
        headers.push(header1);
      } else {
        const header: IMarkdownHeader = {
          id: element.id,
          label: label,
          preview: preview,
          level: elementLevel,
          page: latestH1,
          grouping: latestH1?.label,
        };

        headers.push(header);

        if (elementLevel === 2) {
          tableOfContent[latestH1.id].children.push(header);
        }
      }

      const anchor = this.makeHeaderAnchor(element);
      element.append(anchor);
    }

    // dynamic anchors
    const allElementsWithDynamicAnchor = dom.querySelectorAll(".with-anchor");
    allElementsWithDynamicAnchor.forEach((element) => {
      element.id = kebabCase(element.textContent ?? "");
      const anchor = this.makeHeaderAnchor(element);
      element.append(anchor);
    });

    // dynamic table of content
    const tocElements = dom.querySelectorAll("toc");
    tocElements.forEach((element) => {
      const [, ...pages] = Object.keys(tableOfContent);
      const listItems = pages
        .map((t) => {
          const to = tableOfContent[t];
          return `<li><a href="${to.page.id}">${to.page.label}</a></li>`;
        })
        .join("");
      element.innerHTML = `<ul>${listItems}</ul>`;
    });

    return { dom, tableOfContent, headers };
  }

  private static makeHeaderAnchor(element: Element) {
    const anchor = document.createElement("a");
    anchor.className = "anchor";
    anchor.href = `#${element.id}`;
    return anchor;
  }

  private static isElementHeader(nextElement: Element | null) {
    return !nextElement?.tagName.startsWith("H");
  }

  private static getElementLevel(element: Element) {
    return parseInt(element.tagName.replace("H", ""));
  }
}
