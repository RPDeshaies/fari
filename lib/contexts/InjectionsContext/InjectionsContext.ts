import React from "react";
import { injections } from "../../services/injections";

export const InjectionsContext = React.createContext(injections);
