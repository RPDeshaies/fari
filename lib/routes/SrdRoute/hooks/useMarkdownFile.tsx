import { useEffect, useState } from "react";
import { useLocation } from "react-router";
import { showdownConverter } from "../../../constants/showdownConverter";

export type ILoadFunction = () => Promise<string>;

export type IToc = Record<
  string,
  {
    info: ITocElement;
    level2: Array<ITocElement>;
  }
>;

export type ITocElement = {
  label: string;
  level: number;
  id: string;
};
export function useMarkdownFile(loadFunction: ILoadFunction) {
  const [html, setHtml] = useState<string | undefined>();
  const [toc, setToc] = useState<IToc>({});
  const [dom, setDom] = useState<HTMLDivElement>();
  const location = useLocation();

  useEffect(() => {
    load();
    async function load() {
      if (loadFunction) {
        const markdown = await loadFunction();

        if (markdown) {
          const html = showdownConverter.makeHtml(markdown);
          const dom = document.createElement("div");
          dom.innerHTML = html;

          type IToc = Record<
            string,
            {
              info: ITocElement;
              level2: Array<ITocElement>;
            }
          >;

          let newToc: IToc = {};

          let latestH1Id: string | undefined = undefined;
          dom.querySelectorAll("h1,h2,h3,h4,h5,h6").forEach((e) => {
            // prepare TOC
            const level = parseInt(e.tagName.replace("H", ""));
            if (level === 1) {
              latestH1Id = e.id;
              newToc = {
                ...newToc,
                [e.id]: {
                  info: {
                    id: e.id,
                    label: e.textContent ?? "",
                    level: level,
                  },
                  level2: [],
                },
              };
            }
            if (latestH1Id && level === 2) {
              newToc[latestH1Id].level2.push({
                id: e.id,
                label: e.textContent ?? "",
                level: level,
              });
            }

            // add anchor on header
            const anchor = document.createElement("a");
            anchor.className = "anchor";
            anchor.href = `#${e.id}`;
            e.append(anchor);
          });

          setHtml(dom.innerHTML);
          setToc(newToc);
          setDom(dom);
        }
      }
    }
  }, [loadFunction]);

  useEffect(() => {
    scrollToHeaderOnLoad();
    function scrollToHeaderOnLoad() {
      if (html && location.hash) {
        const element = document.querySelector(location.hash);
        const elementTop = element?.getBoundingClientRect().top ?? 0;
        const topPos = elementTop + window.pageYOffset;
        window.scrollTo({
          top: topPos,
          behavior: "smooth",
        });
      }
    }
  }, [html]);

  return { html, toc, dom };
}
