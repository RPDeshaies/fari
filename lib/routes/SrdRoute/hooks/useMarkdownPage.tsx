import { useMemo } from "react";
import { nextUntil } from "../SrdRoute";

export function useMarkdownPage(
  page: string | undefined,
  dom: HTMLDivElement | undefined
) {
  return useMemo(() => {
    const allH1 =
      dom?.querySelectorAll(`h1`) ?? (([] as unknown) as NodeListOf<Element>);

    const currentH1 = dom?.querySelector(`#${page}`) ?? allH1[0];

    if (!currentH1) {
      return {
        html: "",
        currentH1: allH1[0],
        previousH1: undefined,
        nextH1: allH1[1],
      };
    }

    let previousH1: Element | undefined;
    let nextH1: Element | undefined;

    allH1.forEach((h1, index) => {
      if (h1.id === currentH1.id) {
        previousH1 = allH1[index - 1];
        nextH1 = allH1[index + 1];
      }
    });

    const arrayElement = nextUntil(currentH1, `#${nextH1?.id}`);

    const newDom = document.createElement("div");
    newDom.append(currentH1?.cloneNode(true)!);
    arrayElement.forEach((e) => {
      newDom.append(e.cloneNode(true));
    });

    return { html: newDom?.innerHTML, currentH1, previousH1, nextH1 };
  }, [page, dom]);
}
