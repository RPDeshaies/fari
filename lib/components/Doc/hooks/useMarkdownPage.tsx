import truncate from "lodash/truncate";
import { useMemo } from "react";
import { useLogger } from "../../../contexts/InjectionsContext/hooks/useLogger";

export function useMarkdownPage(options: {
  page: string | undefined;
  section: string | undefined;
  dom: HTMLDivElement | undefined;
}) {
  const { page, dom, section: section } = options;
  const logger = useLogger();

  return useMemo(() => {
    const allH1s =
      dom?.querySelectorAll(`h1`) ?? (([] as unknown) as NodeListOf<Element>);

    if (!!dom && allH1s.length === 0) {
      logger.error("useMarkdownPage: no H1 in markdown document");
    }

    const currentH1 = dom?.querySelector(`[id='${page}']`) ?? allH1s[0];
    const textAfterCurrentH1 = getFirstMatchFromElement(currentH1, "p,ul");
    const currentSectionElement =
      dom?.querySelector(`[id='${section}']`) ?? undefined;

    const textAfterCurrentHash = getFirstMatchFromElement(
      currentSectionElement,
      "p,ul"
    );

    const title =
      currentSectionElement?.textContent ?? currentH1?.textContent ?? "";

    const firstParagraph =
      textAfterCurrentHash?.textContent ??
      textAfterCurrentH1?.textContent ??
      "";
    const description = truncate(firstParagraph, { length: 155 });

    if (!currentH1) {
      return {
        html: "",
        currentH1: allH1s[0],
        previousH1: undefined,
        nextH1: allH1s[1],
        title: title.trim(),
        description: description.trim(),
      };
    }

    let previousH1: Element | undefined;
    let nextH1: Element | undefined;

    allH1s.forEach((h1, index) => {
      if (h1.id === currentH1.id) {
        previousH1 = allH1s[index - 1];
        nextH1 = allH1s[index + 1];
      }
    });

    const allElementsInPage = getAllNextSiblingUntilId(
      currentH1,
      `${nextH1?.id}`
    );
    const newDom = getNewDom(currentH1, allElementsInPage);

    return {
      html: newDom?.innerHTML,
      currentH1,
      previousH1,
      nextH1,
      title: title.trim(),
      description: description.trim(),
    };
  }, [page, dom, section]);
}

function getNewDom(currentH1: Element, allElementsInPage: Element[]) {
  const newDom = document.createElement("div");
  newDom.append(currentH1?.cloneNode(true)!);
  allElementsInPage.forEach((e) => {
    newDom.append(e.cloneNode(true));
  });
  return newDom;
}

function getAllNextSiblingUntilId(
  elem: Element | undefined | null,
  id: string
) {
  return getAllNextSiblingUntilSelector(elem, `[id='${id}']`);
}

function getAllNextSiblingUntilSelector(
  elem: Element | undefined | null,
  selector: string
) {
  const siblings: Array<Element> = [];

  let currentElement = elem?.nextElementSibling;

  while (currentElement) {
    if (currentElement.matches(selector)) break;

    siblings.push(currentElement);

    currentElement = currentElement.nextElementSibling;
  }

  return siblings;
}

function getFirstMatchFromElement(
  elem: Element | undefined | null,
  selector: string
): Element | undefined {
  let currentElement = elem?.nextElementSibling;

  while (currentElement) {
    if (currentElement.matches(selector)) {
      return currentElement;
    }

    currentElement = currentElement.nextElementSibling;
  }
}
