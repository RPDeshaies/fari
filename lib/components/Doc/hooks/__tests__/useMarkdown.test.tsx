/**
 * @jest-environment jsdom
 */
/* eslint-disable react/display-name */

import { renderHook } from "@testing-library/react-hooks";
import React from "react";
import { InjectionsContext } from "../../../../contexts/InjectionsContext/InjectionsContext";
import { useMarkdownFile } from "../useMarkdownFile";
import { useMarkdownPage } from "../useMarkdownPage";

const fakeLogger = {
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
};

beforeEach(() => {
  fakeLogger.debug.mockReset();
  fakeLogger.info.mockReset();
  fakeLogger.warn.mockReset();
  fakeLogger.error.mockReset();
});

describe("useMarkdownFile", () => {
  describe("Given I have an undefined markdown file", () => {
    it("should return an undefined state", async () => {
      const view = renderHook(() => {
        return useMarkdownFile(anUndefinedMarkdownFile, "/test-doc");
      });

      expect(view.result.current.html).toEqual(undefined);
      expect(view.result.current.allHeaders).toEqual([]);
      expect(view.result.current.tableOfContent).toEqual({});
      expect(view.result.current.dom).toEqual(undefined);
    });
  });
  describe("Given I dont have a load function", () => {
    it("should return an undefined state", async () => {
      const view = renderHook(() => {
        return useMarkdownFile(undefined as any, "/test-doc");
      });

      expect(view.result.current.html).toEqual(undefined);
      expect(view.result.current.allHeaders).toEqual([]);
      expect(view.result.current.tableOfContent).toEqual({});
      expect(view.result.current.dom).toEqual(undefined);
    });
  });
  describe("Given I have an empty markdown file", () => {
    it("should return an undefined state", () => {
      const view = renderHook(() => {
        return useMarkdownFile(anEmptyMarkdownFile, "/test-doc");
      });

      expect(view.result.current.html).toEqual(undefined);
      expect(view.result.current.allHeaders).toEqual([]);
      expect(view.result.current.tableOfContent).toEqual({});
      expect(view.result.current.dom).toEqual(undefined);
    });
  });
  describe("Given I have a markdown file without h1", () => {
    it("should return an undefined state", async () => {
      const view = renderHook(
        () => {
          return useMarkdownFile(aMarkdownFileWithoutAHeader1, "/test-doc");
        },
        {
          wrapper: wrapper,
        }
      );

      view.waitFor(() => {
        expect(fakeLogger.error).toHaveBeenCalledWith("");
      });
      expect(view.result.current.html).toEqual(undefined);
      expect(view.result.current.allHeaders).toEqual([]);
      expect(view.result.current.tableOfContent).toEqual({});
      expect(view.result.current.dom).toEqual(undefined);
    });
  });
  describe("Given I have a markdown file with a dynamic anchor tag", () => {
    it("should add an anchor at the tag matching the id", async () => {
      const view = renderHook(
        () => {
          return useMarkdownFile(aMarkdownFileWithADynamicAnchor, "/test-doc");
        },
        {
          wrapper: wrapper,
        }
      );

      await view.waitForValueToChange(() => view.result.current.html);
      expect(view.result.current.html).toMatchSnapshot();
      expect(
        view.result.current.dom
          ?.querySelector("#something-important .anchor")
          ?.getAttribute("href")
      ).toEqual("#something-important");
    });
  });
  describe("Given I have a markdown file with a dynamic table of content element", () => {
    it("should add a table of content", async () => {
      const view = renderHook(
        () => {
          return useMarkdownFile(
            aMarkdownFileWithADynamicTableOfContent,
            "/test-doc"
          );
        },
        {
          wrapper: wrapper,
        }
      );

      await view.waitForValueToChange(() => view.result.current.html);
      expect(view.result.current.dom?.querySelector(".toc")).toMatchSnapshot();
    });
  });
  describe("Given I properly formatted markdown file", () => {
    it("should return an good state", async () => {
      const view = renderHook(() => {
        return useMarkdownFile(aGoodMarkdownFile, "/test-doc");
      });
      await view.waitForValueToChange(() => view.result.current.html);

      expect(view.result.current.html).toMatchSnapshot();
      expect(view.result.current.dom).toMatchSnapshot();
      expect(view.result.current.allHeaders).toEqual([
        {
          grouping: "Page 1",
          id: "page1",
          label: "Page 1",
          level: 1,
          page: undefined,
          preview: "Page 1 introduction",
        },
        {
          grouping: "Page 1",
          id: "header2",
          label: "Header 2",
          level: 2,
          page: {
            grouping: "Page 1",
            id: "page1",
            label: "Page 1",
            level: 1,
            page: undefined,
            preview: "Page 1 introduction",
          },
          preview: "Header 2 details",
        },
        {
          grouping: "Page 2",
          id: "page2",
          label: "Page 2",
          level: 1,
          page: undefined,
          preview: "Page 2 introduction",
        },
        {
          grouping: "Page 3",
          id: "page3",
          label: "Page 3",
          level: 1,
          page: undefined,
          preview: "Page 3 introduction",
        },
        {
          grouping: "Page 3",
          id: "rage",
          label: "Rage",
          level: 2,
          page: {
            grouping: "Page 3",
            id: "page3",
            label: "Page 3",
            level: 1,
            page: undefined,
            preview: "Page 3 introduction",
          },
          preview:
            "Berserk Rage. When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical consequences, you get a free invocation for each. (Fate System Toolkit, p.34)",
        },
      ]);
      expect(view.result.current.tableOfContent).toEqual({
        page1: {
          children: [
            {
              grouping: "Page 1",
              id: "header2",
              label: "Header 2",
              level: 2,
              page: {
                grouping: "Page 1",
                id: "page1",
                label: "Page 1",
                level: 1,
                page: undefined,
                preview: "Page 1 introduction",
              },
              preview: "Header 2 details",
            },
          ],
          page: {
            grouping: "Page 1",
            id: "page1",
            label: "Page 1",
            level: 1,
            page: undefined,
            preview: "Page 1 introduction",
          },
        },
        page2: {
          children: [],
          page: {
            grouping: "Page 2",
            id: "page2",
            label: "Page 2",
            level: 1,
            page: undefined,
            preview: "Page 2 introduction",
          },
        },
        page3: {
          children: [
            {
              grouping: "Page 3",
              id: "rage",
              label: "Rage",
              level: 2,
              page: {
                grouping: "Page 3",
                id: "page3",
                label: "Page 3",
                level: 1,
                page: undefined,
                preview: "Page 3 introduction",
              },
              preview:
                "Berserk Rage. When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical consequences, you get a free invocation for each. (Fate System Toolkit, p.34)",
            },
          ],
          page: {
            grouping: "Page 3",
            id: "page3",
            label: "Page 3",
            level: 1,
            page: undefined,
            preview: "Page 3 introduction",
          },
        },
      });
    });
  });
});

describe("useMarkdownPage", () => {
  describe("Given an empty page", () => {
    it("should default to first h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(anEmptyMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: "", section: "", dom: dom });
      });

      expect(view.result.current.title).toEqual("");
      expect(view.result.current.description).toEqual("");
      expect(view.result.current.currentH1).toEqual(undefined);
      expect(view.result.current.previousH1).toEqual(undefined);
      expect(view.result.current.nextH1).toEqual(undefined);
    });
  });
  describe("Given an undefined page", () => {
    it("should default to first h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(anUndefinedMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: undefined, section: "", dom: dom });
      });

      expect(view.result.current.title).toEqual("");
      expect(view.result.current.description).toEqual("");
      expect(view.result.current.currentH1).toEqual(undefined);
      expect(view.result.current.previousH1).toEqual(undefined);
      expect(view.result.current.nextH1).toEqual(undefined);
    });
  });
  describe("Given an undefined page", () => {
    it("should default to first h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(aGoodMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: undefined, section: "", dom: dom });
      });

      await view.waitForNextUpdate();
      expect(view.result.current.title).toEqual("Page 1");
      expect(view.result.current.description).toEqual("Page 1 introduction");
      expect(view.result.current.currentH1.textContent).toEqual("Page 1");
      expect(view.result.current.previousH1).toEqual(undefined);
      expect(view.result.current.nextH1?.textContent).toEqual("Page 2");
    });
  });
  describe("Given the first page page", () => {
    it("should go to first h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(aGoodMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: undefined, section: "", dom: dom });
      });

      await view.waitForNextUpdate();
      expect(view.result.current.title).toEqual("Page 1");
      expect(view.result.current.description).toEqual("Page 1 introduction");
      expect(view.result.current.currentH1.textContent).toEqual("Page 1");
      expect(view.result.current.previousH1).toEqual(undefined);
      expect(view.result.current.nextH1?.textContent).toEqual("Page 2");
    });
  });
  describe("Given the second page", () => {
    it("should go to second h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(aGoodMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: "page2", section: "", dom });
      });

      await view.waitForNextUpdate();
      expect(view.result.current.title).toEqual("Page 2");
      expect(view.result.current.description).toEqual("Page 2 introduction");
      expect(view.result.current.currentH1.textContent).toEqual("Page 2");
      expect(view.result.current.previousH1?.textContent).toEqual("Page 1");
      expect(view.result.current.nextH1?.textContent).toEqual("Page 3");
    });
  });
  describe("Given the third page", () => {
    it("should go to third h1", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(aGoodMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: "page3", section: "", dom });
      });

      await view.waitForNextUpdate();
      expect(view.result.current.title).toEqual("Page 3");
      expect(view.result.current.description).toEqual("Page 3 introduction");
      expect(view.result.current.currentH1.textContent).toEqual("Page 3");
      expect(view.result.current.previousH1?.textContent).toEqual("Page 2");
      expect(view.result.current.nextH1).toEqual(undefined);
    });
  });
  describe("Given the third page and there is a hash", () => {
    it("should go to third h1 and have the right title and description", async () => {
      const view = renderHook(() => {
        const { dom } = useMarkdownFile(aGoodMarkdownFile, "/test-doc");

        return useMarkdownPage({ page: "page3", section: "rage", dom });
      });

      await view.waitForNextUpdate();
      expect(view.result.current.title).toEqual("Rage");
      expect(view.result.current.description).toEqual(
        "Berserk Rage. When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical con..."
      );
      expect(view.result.current.currentH1.textContent).toEqual("Page 3");
      expect(view.result.current.previousH1?.textContent).toEqual("Page 2");
      expect(view.result.current.nextH1).toEqual(undefined);
    });
  });
});

const wrapper = (props: { children?: React.ReactNode }): JSX.Element => {
  return (
    <InjectionsContext.Provider value={{ logger: fakeLogger } as any}>
      {props.children}
    </InjectionsContext.Provider>
  );
};

function makeLoadFunction(markdownFile: string | undefined) {
  return async () => new Promise<string>((r) => r(markdownFile as string));
}

const anUndefinedMarkdownFile = makeLoadFunction(undefined);
const anEmptyMarkdownFile = makeLoadFunction("");
const aGoodMarkdownFile = makeLoadFunction(`
# Page 1

Page 1 introduction

## Header 2

Header 2 details

# Page 2

Page 2 introduction

# Page 3

Page 3 introduction

- bullet 1
- bullet 2


## Rage

- **Berserk Rage**. When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical consequences, you get a free invocation for each. (Fate System Toolkit, p.34)

`);

const aMarkdownFileWithoutAHeader1 = makeLoadFunction(`
## Header 2

Header 2 details
`);

const aMarkdownFileWithADynamicAnchor = makeLoadFunction(`
# Header 1

<p class="with-anchor">something important</p>
`);

const aMarkdownFileWithADynamicTableOfContent = makeLoadFunction(`
# 1

<toc/>

## 1.1

## 1.2

## 1.3

# 2

## 2.1
## 2.2
## 2.3

# 3

# 4

`);
