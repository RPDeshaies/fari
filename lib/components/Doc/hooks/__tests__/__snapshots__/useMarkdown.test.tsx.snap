// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMarkdownFile Given I have a markdown file with a dynamic anchor tag should add an anchor at the tag matching the id 1`] = `
"<h1 id=\\"header1\\">Header 1<a class=\\"anchor\\" href=\\"#header1\\"></a></h1>
<p class=\\"with-anchor\\" id=\\"something-important\\">something important<a class=\\"anchor\\" href=\\"#something-important\\"></a></p>"
`;

exports[`useMarkdownFile Given I have a markdown file with a dynamic table of content element should add a table of content 1`] = `
<ul
  class="toc"
>
  <li>
    <a
      href="2"
    >
      2
    </a>
  </li>
  <li>
    <a
      href="3"
    >
      3
    </a>
  </li>
  <li>
    <a
      href="4"
    >
      4
    </a>
  </li>
</ul>
`;

exports[`useMarkdownFile Given I properly formatted markdown file should return an good state 1`] = `
"<h1 id=\\"page1\\">Page 1<a class=\\"anchor\\" href=\\"#page1\\"></a></h1>
<p>Page 1 introduction</p>
<h2 id=\\"header2\\">Header 2<a class=\\"anchor\\" href=\\"#header2\\"></a></h2>
<p>Header 2 details</p>
<h1 id=\\"page2\\">Page 2<a class=\\"anchor\\" href=\\"#page2\\"></a></h1>
<p>Page 2 introduction</p>
<h1 id=\\"page3\\">Page 3<a class=\\"anchor\\" href=\\"#page3\\"></a></h1>
<p>Page 3 introduction</p>
<ul>
<li>bullet 1</li>
<li>bullet 2</li>
</ul>
<h2 id=\\"rage\\">Rage<a class=\\"anchor\\" href=\\"#rage\\"></a></h2>
<ul>
<li><strong>Berserk Rage</strong>. When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical consequences, you get a free invocation for each. (Fate System Toolkit, p.34)</li>
</ul>"
`;

exports[`useMarkdownFile Given I properly formatted markdown file should return an good state 2`] = `
<div>
  <h1
    id="page1"
  >
    Page 1
    <a
      class="anchor"
      href="#page1"
    />
  </h1>
  

  <p>
    Page 1 introduction
  </p>
  

  <h2
    id="header2"
  >
    Header 2
    <a
      class="anchor"
      href="#header2"
    />
  </h2>
  

  <p>
    Header 2 details
  </p>
  

  <h1
    id="page2"
  >
    Page 2
    <a
      class="anchor"
      href="#page2"
    />
  </h1>
  

  <p>
    Page 2 introduction
  </p>
  

  <h1
    id="page3"
  >
    Page 3
    <a
      class="anchor"
      href="#page3"
    />
  </h1>
  

  <p>
    Page 3 introduction
  </p>
  

  <ul>
    

    <li>
      bullet 1
    </li>
    

    <li>
      bullet 2
    </li>
    

  </ul>
  

  <h2
    id="rage"
  >
    Rage
    <a
      class="anchor"
      href="#rage"
    />
  </h2>
  

  <ul>
    

    <li>
      <strong>
        Berserk Rage
      </strong>
      . When you suffer a physical consequence, you can invoke that consequence for free on your next attack. If you suffer multiple physical consequences, you get a free invocation for each. (Fate System Toolkit, p.34)
    </li>
    

  </ul>
</div>
`;
